File: ios/LiveUpdatesProvider/Sources/LiveUpdatesProvider/LiveUpdatesRegistry.swift
import Foundation

public final class LiveUpdatesRegistry: @unchecked Sendable {
    public static let shared = LiveUpdatesRegistry()
    
    private var providers = [String: any LiveUpdatesProvider]()
    
    private let lock = NSLock()
    
    private init() {}
    
    public func register(_ provider: any LiveUpdatesProvider) {
        lock.lock()
        defer { lock.unlock() }
        
        if providers[provider.id] != nil {
            let message = "LiveUpdatesRegistry: Provider with ID '\(provider.id)' is already registered. Ignoring subsequent registration."
            print(message)
            assertionFailure(message)
            return
        }
        
        providers[provider.id] = provider
    }
    
    public func resolve(_ id: String) -> (any LiveUpdatesProvider)? {
        lock.lock()
        defer { lock.unlock() }
        return providers[id]
    }
    
    public func require(_ id: String) throws -> any LiveUpdatesProvider {
        guard let provider = resolve(id) else {
            throw LiveUpdatesError.providerNotRegistered(id)
        }
        return provider
    }
}

File: android/live-updates-provider/src/main/kotlin/io/ionic/liveupdatesprovider/LiveUpdatesRegistry.kt
package io.ionic.liveupdatesprovider

import android.util.Log
import java.util.concurrent.ConcurrentHashMap


object LiveUpdatesRegistry {
    private val providers: ConcurrentHashMap<String, LiveUpdatesProvider> = ConcurrentHashMap()


    @JvmStatic
    fun register(provider: LiveUpdatesProvider) {
        if (provider.id.isBlank()) {
            Log.w("LiveUpdatesRegistry", "Cannot register a provider with an empty ID")
            return
        }
        val previous = providers.putIfAbsent(provider.id, provider)
        if (previous != null) {
            Log.w("LiveUpdatesRegistry", "Provider with ID '${provider.id}' is already registered. Ignoring subsequent registration.")
            return
        }
    }

    @JvmStatic
    fun resolve(providerId: String): LiveUpdatesProvider? {
        return providers[providerId]
    }

    @JvmStatic
    @Throws(LiveUpdatesError.ProviderNotRegistered::class)
    fun require(providerId: String): LiveUpdatesProvider {
        return resolve(providerId)
            ?: throw LiveUpdatesError.ProviderNotRegistered(providerId)
    }

    internal fun clear() {
        providers.clear()
    }
}
